### PR TITLE
feat(simperium) new type definitions

### PR DESCRIPTION
**Pull request by dmsnell**
Add type definitions and stub test file for `simperium` library
https://github.com/automattic/node-simperium

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

Adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` should be present and it shouldn't have any additional or disabling of rules. Just content as `{ "extends": "dtslint/dt.json" }`. If for reason the some rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]`  and not for whole package so that the need for disabling can be reviewed.
- [x] `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

---

This is my first submission to DefinitelyTyped. I tried to follow all the rules and steps but I appreciate your kindness if I overlooked any or didn't understand them properly. Thanks!

### CODE REVIEW

File: types/simperium/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,228 @@
+// Type definitions for simperium 1.1
+// Project: https://github.com/Simperium/node-simperium#readme
+// Definitions by: Dennis Snell <https://github.com/me>
```
Comment (danger-public) [inline types/simperium/index.d.ts:3]: - :no_entry_sign: This line should have your github username in it, not /me

**Comment by danger-public** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## simperium ([<kbd>unpkg</kbd>](https://unpkg.com/browse/simperium@latest/))
was missing the following properties:
1. Client
1. Bucket
1. util
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 19eb8fcc215dbf8fa21b6ae81737c30d392dca30
</p>

**Comment by dmsnell** [comment]
> What I'm concerned about is which API should be exported directly

Thanks for the review @peterblazejewicz!!!

I'm not sure how to answer though. I'm trying to mix being descriptive and proscriptive in here. For that reason I've renamed some parameters while exporting more than may be necessary (because some people use the internals when they shouldn't).

I didn't realize the parameter naming would be important since it doesn't have any impact on the built code or types of the parameters. This library has some legacy issues with naming (unrelated to `createClient` vs. `initClient` which is more of a happen-stance change).

Thanks for all your help. If it matters I don't think many people are using this library outside of ourselves. It's open-source and should be but we're also our major producer and consumer at the same time, if that makes sense 🙃

**Comment by peterblazejewicz** [comment]
I'm only concerned about those differences, this is to be published to NPM as the shape reflecting usage of the package (with from Flow to TS migration). I understand your points here.
I'm concerned about reading README of the origal package 'as is'. I cannot see the `initClient` function defined anywhere:
https://github.com/Simperium/node-simperium/search?q=initClient&unscoped_q=initClient
or for example, looking into the sample code, it won't compile with those types:
```ts
/// <reference types="node" />
import { Auth, Client } from 'simperium';
// OR
// import Client from 'simperium/client';
// import { Auth } from 'simperium/auth';

const client = new Client(process.env.SIMPERIUM_APP_ID!);
const auth = new Auth(process.env.SIMPERIUM_APP_ID!, process.env.SIMPERIUM_APP_SECRET!);

let interval = Number.POSITIVE_INFINITY;
let lastInt: Date;

client.connect();

auth.on('authorize', user => {
    const notes = client.bucket('note');

    client.accessToken = user.access_token;

    notes.on('index', () => {
        const now = new Date().getTime();
        const note = {
            content: 'Hola mundo!',
            systemTags: [],
            tags: [],
            creationDate: now,
            modificationDate: now,
            publishURL: '',
            shareURL: '',
            deleted: false,
        };

        notes.add(note, (err, id) => {
            setTimeout(() => {
                note.content += '\n\nEl Fin';
                note.deleted = true;
                notes.update(id, note);
                notes.remove(id);
            }, 2000);
        });
    });
});

auth.authorize(process.env.SIMPERIUM_USERNAME!, process.env.SIMPERIUM_PASSWORD!).then(
    user => {
        console.log('Logged in', user);
    },
    error => {
        console.error('Failed to authenticate', error, error.stack);
        throw error;
    },
);

client.on('send', data => {
    console.warn(' => ', data.slice(0, 200));
});

client.on('message', message => {
    console.warn(' <= ', message.slice(0, 200));
});

client.on('reconnect', attempt => {
    console.warn('Attempting reconnection', attempt);
});

process.on('SIGINT', () => {
    if (lastInt) {
        interval = new Date().getTime() - lastInt.getTime();
    }

    if (interval < 500) {
        console.log('Shutting down');
        process.exit();
    }

    lastInt = new Date();
    console.log('Reconnecting, press interrupt again to exit');
    console.log('Disconnect and reconnect');
    client.disconnect();
});
```
Above is just taken 'as is' from project sample indicated in the README.

If it is OK for you, can you consider adding the README.md to the PR?. It will be published with the package to the @types/... If someone has a doubt how to use api, the README is the way to provide that insights.
For example:
https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/react-autosuggest
This is only my personal opinion. I'd be more than happy to review.

**Comment by dmsnell** [comment]
@peterblazejewicz that makes sense and I'll get rid of `initClient` - it's not that important to hold that name. can also add the readme and review everything.

thank you so much for your help. I didn't realize I'd get such awesome attention on this and I'm super grateful.

it's likely I won't get any more updates until early or mid next-week. 🙇‍♂️

**Comment by elibarzilay** [comment]
@dmsnell: ping?

**Comment by dmsnell** [comment]
@peterblazejewicz I reverted the name change and added a README
@elibarzilay thanks for the ping! - pong!

**Pull request merged**